### PR TITLE
Adding initial API for custom IAM profile

### PR DESCRIPTION
**Pull request by chrislovecnm**
If we are going to have Security Groups on the IG, then the IAM profile would make sense on the IG as well.

@robinpercy @justinsb @KashifSaadat @gambol99 @blakebarnett  thoughts?

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Yes I know this is a partial API, well complete once we have an accord.

**Comment by justinsb** [comment]
This makes sense to me.  I like having this per IG.  We do have additionalPolicies in the cluster spec, but I hope that will at some stage be deprecated.

One option is to consider building an `iam` type, and having `name` as a field in it, because this gives us room to expand.  But in the absence of a use-case, we can do that when we need it.  (It's a relatively easy conversion)

**Comment by justinsb** [comment]
@chrislovecnm and I discussed on slack that it was probably a good idea to go straight to the object:

i.e.

```
spec:
  iam:
    profile: <name>
```

Because we expect that people will want more than just the profile name...

**Comment by chrislovecnm** [comment]
On the instancegroupspex

**Comment by chrislovecnm** [comment]
@justinsb named challenged again.  We have `IAMSpec` in `ClusterSpec`.   I added the api as we agreed upon.  Let me know what the heck we can call this.

Any thoughts peeps what to call this internally?

**Comment by KashifSaadat** [comment]
@chrislovecnm you could maybe incorporate it within the ClusterSpec `IAMSpec` and extend the `IAMProfile` to take a list of InstanceGroups / InstanceGroupRoles to apply for? That would avoid the similar naming.

**Comment by chrislovecnm** [comment]
I hear what you are saying, but I think that would be a bit messy as well :(

File: pkg/apis/kops/instancegroup.go [hunk 1]
```diff
@@ -109,6 +109,16 @@ type InstanceGroupSpec struct {
 	Kubelet *KubeletConfigSpec `json:"kubelet,omitempty"`
 	// Taints indicates the kubernetes taints for nodes in this group
 	Taints []string `json:"taints,omitempty"`
+	// IAMProfileSpec defines the identity of the cloud group iam profile.
+	IAM *IAMProfileSpec `json:"iam,omitempty"`
```
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: How about `InstanceGroupIAMSpec` ?  It isn't particularly visible to users, though it is visible to code clients etc.

File: pkg/apis/kops/instancegroup.go [hunk 2]
```diff
@@ -109,6 +109,16 @@ type InstanceGroupSpec struct {
...
+}
+
+// IAMProfileSpec defined the identity of the cloud group iam profile.
```
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: Present tense I think i.e. defines, not defined.

But how about "IAMProfileSpec contains the IAM related settings for the instance group"

File: pkg/apis/kops/instancegroup.go [hunk 3]
```diff
@@ -109,6 +109,16 @@ type InstanceGroupSpec struct {
 	Kubelet *KubeletConfigSpec `json:"kubelet,omitempty"`
 	// Taints indicates the kubernetes taints for nodes in this group
 	Taints []string `json:"taints,omitempty"`
+	// IAMProfileSpec defines the identity of the cloud group iam profile.
```
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: cloud group isn't really an API concept, so probably doesn't belong here.

File: pkg/apis/kops/instancegroup.go [hunk 4]
```diff
@@ -109,6 +109,16 @@ type InstanceGroupSpec struct {
...
+	IAM *IAMProfileSpec `json:"iam,omitempty"`
+}
+
+// IAMProfileSpec defined the identity of the cloud group iam profile.
+// In EC2 the cloud group is defined by the launch configuration.
+type IAMProfileSpec struct {
+	// Profile of the cloud group iam profile. In aws this is the arn
+	// for the iam instance profile
+	Profile *string `json:"profile,omitempty"`
```
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: How about `Profile is the AWS IAM Profile to attach to instances in this instance group.  Specify the ARN for the IAM instance profile.  (AWS only)`

**Comment by chrislovecnm** [comment]
/hold

@rdrgmnzs you want to take a look?

**Comment by justinsb** [comment]
We merged #4171 which I believe incorporates this PR.  Closing - please reopen if that's not righ!

**Pull request closed without merging**